pkg/validationfile/blocks.go

```diff
@@ -0,0 +1,444 @@
+		trimmed := strings.TrimSpace(assertion.relationshipString)
+		parsed := tuple.Parse(trimmed)
+		if parsed == nil {
+			return relationships, &ErrorWithSource{
+				fmt.Errorf("could not parse relationship `%s`", assertion.relationshipString),
+				assertion.relationshipString,
+				uint32(assertion.lineNumber),
+				uint32(assertion.columnPosition),
+			}
+		}
+		relationships = append(relationships, ParsedAssertion{
+			Relationship:   parsed,
+			LineNumber:     uint32(assertion.lineNumber),
+			ColumnPosition: uint32(assertion.columnPosition),
+		})
+	}
+	return relationships, nil
+}
+
+// AssertFalseRelationships returns the relationships for which to assert
+// non-existence.
+func (a *Assertions) AssertFalseRelationships() ([]ParsedAssertion, *ErrorWithSource) {
+	if a == nil {
+		return nil, nil
+	}
+
+	relationships := make([]ParsedAssertion, 0, len(a.AssertFalse))
+	for _, assertion := range a.AssertFalse {
+		trimmed := strings.TrimSpace(assertion.relationshipString)
+		parsed := tuple.Parse(trimmed)
+		if parsed == nil {
+			return relationships, &ErrorWithSource{
+				fmt.Errorf("could not parse relationship `%s`", assertion.relationshipString),
+				assertion.relationshipString,
+				uint32(assertion.lineNumber),
+				uint32(assertion.columnPosition),
+			}
+		}
+		relationships = append(relationships, ParsedAssertion{
+			Relationship:   parsed,
+			LineNumber:     uint32(assertion.lineNumber),
+			ColumnPosition: uint32(assertion.columnPosition),
+		})
+	}
+	return relationships, nil
+}
+
+// ParseRelationshipsBlock parses a block of newline-separated relationships into a set
+// of relation tuples.
+func ParseRelationshipsBlock(contents []byte) ([]*v1.Relationship, error) {
+	// Unmarshal to a node, so we can get line and col information.
+	var node yamlv3.Node
+	err := yamlv3.Unmarshal(contents, &node)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(node.Content) == 0 {
+		return nil, nil
+	}
+
+	if node.Content[0].Kind != yamlv3.ScalarNode {
+		return nil, fmt.Errorf("expected string at top level")
+	}
+
+	return DecodeRelationshipsBlock(node.Value, node.Line)
+}
+
+// DecodeRelationshipsBlock decodes a YAML node containing a block of newline-separated
+// relationships into a set of relation tuples.
+func DecodeRelationshipsBlock(value string, startingLineIndex int) ([]*v1.Relationship, error) {
+	parsed, err := ParseRelationships(value)
+	if err != nil {
+		var errWithSource ErrorWithSource
+		if errors.As(err, &errWithSource) {
+			return nil, ErrorWithSource{
+				errWithSource.error,
+				errWithSource.Source,
+				errWithSource.LineNumber + uint32(startingLineIndex),
+				errWithSource.ColumnPosition,
+			}
+		}
+	}
+	return parsed, err
+}
+
+// ParseRelationships parses a newline-separated relationships string into a set
+// of relation tuples.
+func ParseRelationships(relationshipsString string) ([]*v1.Relationship, error) {
+	if relationshipsString == "" {
+		return nil, nil
```

Comment:
this is kind of a nit but I think I would find it surprising that this doesn't either return an error (no relationships) or an empty slice